---
title: "lab4"
output: html_document
---

```{bash, eval = FALSE}
wget -nc https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/735/GCF_000001735.4_TAIR10.1/GCF_000001735.4_TAIR10.1_genomic.fna.gz

#using gunzip to open
gunzip GCF_000001735.4_TAIR10.1_genomic.fna.gz

#GFF file with the gene annotation
wget -nc https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/735/GCF_000001735.4_TAIR10.1/GCF_000001735.4_TAIR10.1_genomic.gtf.gz

gunzip GCF_000001735.4_TAIR10.1_genomic.gtf.gz
```


```{r, eval = FALSE}
#Items to downlaod 
BiocManager::install("EnhancedVolcano")
BiocManager::install("clusterProfiler")

lapply(c(
  "docopt", "DT", "pheatmap","GenomicFeatures", "DESeq2",
  "edgeR", "systemPipeR", "systemPipeRdata", "BiocStyle", "GO.db", "dplyr",
  "tidyr", "stringr", "Rqc", "QuasR", "DT", "ape", "clusterProfiler","biomaRt","EnhancedVolcano"
), require, character.only = TRUE)

```

#q1
```{r load data, echo=TRUE, warning=FALSE}
library(systemPipeRdata)
library(DT)

# Get the path to the directory with the raw FASTQ files
fq_path <- systemPipeRdata::pathList()$fastqdir

# Load the original, clean targets file directly from the package
meta_file_path <- system.file("extdata", "param", "targetsPE.txt", package = "systemPipeRdata")
meta_data <- read.table(meta_file_path, header = TRUE)

# Join the directory and file names
meta_data$FileName1 <- gsub('./data/', fq_path, meta_data$FileName1)
meta_data$FileName2 <- gsub('./data/', fq_path, meta_data$FileName2)

DT::datatable(meta_data, options = list(pageLength = 10))
```

#q2
```{r plots,echo=TRUE, warning=FALSE, message=FALSE}
library(Rqc)
qa_report <- rqc(path = fq_path, pattern = "\\.fastq\\.gz$")

# per-cycle Q-score boxplots
rqcCycleQualityBoxPlot(qa_report[1:12]) 
rqcCycleQualityBoxPlot(qa_report[13:24]) 
rqcCycleQualityBoxPlot(qa_report[25:36])

# per-cycle base call frequency plts
rqcCycleBaseCallsLinePlot(qa_report[1:12]) 
rqcCycleBaseCallsLinePlot(qa_report[13:24]) 
rqcCycleBaseCallsLinePlot(qa_report[25:36]) + ggtitle("Base Calls: Files 25-36")
```

#q3
```{r process_reads_final, eval = FALSE}
library(QuasR)
library(DT)

#directory
if (!dir.exists("processed_reads")) {
  dir.create("processed_reads")
}

for (i in seq_along(meta_data$FileName1)) {
  
  current_file <- meta_data$FileName1[i]
  
  base_name <- tools::file_path_sans_ext(basename(current_file), compression = TRUE)
  new_filename <- paste0(base_name, "_processed.fastq.gz")
  output_file <- file.path("processed_reads", new_filename)

  QuasR::preprocessReads(filename = current_file,
                         outputFilename = output_file,
                         truncateEndBases = 3,
                         Rpattern = 'GCCCGGGTAA',
                         nBases = 1) 

  meta_data$FileName1[i] <- output_file
}


# Finally, display the correctly updated metadata table
DT::datatable(meta_data)
```


#q4
```{bash terminal, eval= FALSE}
#to unzip file
cd processed_reads
(base) T450-XXXXXXX:processed_readsgunzip *.gz
```

```{r directories, eval= FALSE}
#run this once for the directories 
dir.create("sam_files", showWarnings = FALSE)
dir.create("bam_files", showWarnings = FALSE)
writeLines(c("sam_files/", "bam_files/"), con = ".gitignore")
```

```{r align_reads_simple, eval= FALSE, message=TRUE, warning=TRUE}
hisat2_index <- "./hisat2_index/tair10_1_index"

for(i in seq_along(meta_data$FileName1)){
  
  
  sample_name <- meta_data$SampleName[i]
  #print(sample_name)
  fastq_file <- gsub(".gz$", "", meta_data$FileName1[i])
  #print(fastq_file)
  
  sam_file <- file.path("sam_files", paste0(sample_name, ".sam"))
  bam_file <- file.path("bam_files", paste0(sample_name, ".bam"))
  sorted_bam_file <- file.path("bam_files", paste0(sample_name, "_sorted.bam"))
  log_file <- file.path("bam_files", paste0(sample_name, "_hisat2.log"))
  
  # is my file missing
  if (!file.exists(fastq_file)) {
    message("Skipping ", sample_name, ": FASTQ file not found.")
    next # 'next' jumps to the next iteration of the loop
  }
  
  # run HISAT2 
  hisat2_log <- tryCatch({
    system2("hisat2", args = c("-p", "8",
                              "-x", hisat2_index,
                              "-U", fastq_file,
                              "-S", sam_file),
            stdout = TRUE, stderr = TRUE)
  }, error = function(e) paste("HISAT2 failed:", e$message))
  
  # write output from hisat2 to its log file
  writeLines(hisat2_log, log_file)
  
  # run sam tools 
  tryCatch({
    # 1. Convert SAM to BAM
    system2("samtools", args = c("view", "-bS", sam_file, "-o", bam_file))
    # 2. Sort the BAM file
    system2("samtools", args = c("sort", bam_file, "-o", sorted_bam_file))
    # 3. Index the sorted BAM file
    system2("samtools", args = c("index", sorted_bam_file))
  }, error = function(e) {
    message("Samtools failed for ", sample_name, ": ", e$message)
  })
}

print("complete.")
```

#q5
```{r idos code, eval=FALSE} 
# Copy and paste from idos lab 

# get the folder
hisat2_logs_dir <- "bam_files/" #changed this to my directory

# 1. Get a list of all HISAT2 log files
log_files <- list.files(hisat2_logs_dir, pattern = "*.log", full.names = TRUE) #changed this to *.log

#preper an "empty" vector of the apropreate size, faster than appending. 
percent_aligned <- 1:length(log_files)

#loop through log files
for (i in seq_along(percent_aligned)) {
  
  percent_aligned[i] <- readLines(log_files[i])[length(readLines(log_files[i]))]
  
}

#bind vectors as dataframe
align_df <- data.frame(sort(meta_data$SampleName),percent_aligned)

# extract the numeric percent value
align_df <- align_df %>% 
  mutate(percent_aligned = as.numeric(
    stringr::str_split_i(align_df$percent_aligned, "%",1))) %>% 
  rename(sample_name = sort.meta_data.SampleName.)

head(align_df)
```
### **--------- Question 5 ---------**


```{r alignment_stats, echo=TRUE, message=FALSE, warning=FALSE}
library(dplyr)
library(stringr)
library(DT)
library(ggplot2)


log_dir <- "bam_files/"

log_files <- list.files(log_dir, pattern = "_hisat2.log", full.names = TRUE)


sample_names <- str_replace(basename(log_files), "_hisat2.log", "")

alignment_rates <- sapply(log_files, function(file) {
  # Read the file once
  log_content <- readLines(file)
  # The alignment rate is always the last line of the hisat2 summary
  summary_line <- log_content[length(log_content)]
  # Extract just the numeric percentage and convert it to a number
  as.numeric(str_extract(summary_line, "\\d+\\.\\d+"))
})

align_df <- data.frame(
  SampleName = sample_names,
  PercentAligned = alignment_rates
)


datatable(align_df,
          caption = "Overall Alignment Rate per Sample",
          options = list(pageLength = 10, autoWidth = TRUE),
          rownames = FALSE)


ggplot(align_df, aes(y = PercentAligned)) +
  geom_boxplot(fill = "pink", color = "maroon", outlier.colour = "blue") +
  # Add individual points over the boxplot for better visibility
  labs(title = "Box Plot of Alignment Percentages",
       y = "Percent Aligned (%)",
       x = "") +
  theme_minimal(base_size = 14) +
  # Remove the x-axis text and ticks as they are not needed for a single boxplot
  theme(axis.text.x = element_blank(),
        axis.ticks.x = element_blank())
```

#q6
```{r Install Rsubred, eval = FALSE } 
BiocManager::install("Rsubread")
```


```{r copy+paste from idos lab, eval = FALSE}
#list bam files
bfiles <- list.files("bam_files/", pattern = "_sorted.bam$", full.names = TRUE) #directoreis slightly changed 

#Counting how many reads corespond to each gene
gene_count_list <- Rsubread::featureCounts(files = bfiles, annot.ext = "GCF_000001735.4_TAIR10.1_genomic.gtf", 
                        isGTFAnnotationFile = TRUE, # <--- input annotation is GTF
                        allowMultiOverlap = FALSE,  # <--- don't allow reads that overlap with multiple loci
                        isPairedEnd = FALSE, nthreads = 8,
                        minMQS = 10, # <--- minimum mapping quality score of 10 (like a phred score for the hisat2 alignment)
                        GTF.featureType = "exon",  # <--- Count reads overlapping 'exon' features
                        GTF.attrType = "gene_id" # <--- Groups exon by gene id
                        )
```

### **--------- Question 6 ---------**

```{r create_count_table, echo=TRUE, message=FALSE, warning=FALSE}
library(Rsubread)
library(DT)

bam_files <- list.files("./bam_files", pattern = "_sorted.bam$", full.names = TRUE)
gtf_file <- "GCF_000001735.4_TAIR10.1_genomic.gtf"

gene_counts <- featureCounts(
  files = bam_files,
  annot.ext = gtf_file,
  isGTFAnnotationFile = TRUE,
  isPairedEnd = FALSE, # FALSE because alignment was done in single-end mode
  GTF.featureType = "exon",    #count reads overlapping exons
  GTF.attrType = "gene_id",    #group counts by gene ID
  nthreads = 8                 #ise 8 threads to speed up counting
)


count_table <- gene_counts$counts
colnames(count_table) <- gsub("_sorted.bam", "", colnames(count_table))# remove the "_sorted.bam" suffix


count_table <- count_table[rowSums(count_table) > 0, ]


DT::datatable(count_table,
          caption = "Gene Read Counts per Sample (Filtered)",
          options = list(pageLength = 10, autoWidth = TRUE))
```

#q7
```{r create_deseq_objects, message=FALSE, warning=FALSE}
library(DESeq2)
filter_and_count <- function(res_obj, comparison_name, fc_threshold = 2) {
  # Remove NAs
  res_filtered <- res_obj[!is.na(res_obj$padj) & !is.na(res_obj$log2FoldChange), ]
  
  # Apply filters: |log2FC| >= log2(2) = 1 and padj <= alpha (already set in results())
  sig_genes <- res_filtered[abs(res_filtered$log2FoldChange) >= log2(fc_threshold), ]
  
  # Count up and down regulated
  up_regulated <- sum(sig_genes$log2FoldChange > 0)
  down_regulated <- sum(sig_genes$log2FoldChange < 0)
  
  return(data.frame(
    Comparison = comparison_name,
    Up_regulated = up_regulated,
    Down_regulated = down_regulated
  ))
}
coldata <- meta_data %>% 
  dplyr::select(SampleName, SampleLong, Factor) %>% 
  dplyr::mutate(SampleLong = str_split_i(SampleLong, "\\.", 1)) %>%
  dplyr::rename(condition = SampleLong) %>%
  dplyr::mutate(condition = factor(condition),
                Factor = factor(Factor),
                SampleName = factor(SampleName))

rownames(coldata) <- coldata$SampleName
coldata$type <- factor(rep("single-read", nrow(coldata)))

coldata <- coldata[match(colnames(count_table), rownames(coldata)), ]

# Final check that the order is correct
all(rownames(coldata) == colnames(count_table))


# combines your raw count_table with the coldata into a single object.
dds2 <- DESeqDataSetFromMatrix(countData = count_table,
                               colData = coldata,
                               design = ~ Factor)


# -run the DESeq Analysis ---
# This performs the actual differential expression analysis
dds2_results <- DESeq(dds2)
```

### **--------- Question 7 ---------**

```{r final_deseq2_analysis, echo=TRUE, message=FALSE, warning=FALSE}
targets_path <- system.file("extdata", "param", "targetsPE.txt", package = "systemPipeRdata")
comp <- systemPipeR::readComp(file = targets_path, format = "matrix", delim = "-")
comparison_pairs <- comp[[1]]

results_list <- list()

for (i in 1:nrow(comparison_pairs)) {
  
  group1 <- comparison_pairs[i, 1]
  group2 <- comparison_pairs[i, 2]
  
  comparison_name <- paste(group1, "vs", group2)
  
  res <- DESeq2::results(dds2_results, contrast = c("Factor", group1, group2), alpha = 0.2)
  
  filtered_results <- filter_and_count(res, comparison_name)
  
  results_list[[comparison_name]] <- filtered_results
}

de_summary_by_factor <- do.call(rbind, results_list)

plot_data <- de_summary_by_factor %>%
  tidyr::pivot_longer(cols = c(Up_regulated, Down_regulated),
                      names_to = "Regulation",
                      values_to = "Count")

ggplot(plot_data, aes(x = Comparison, y = Count, fill = Regulation)) +
  geom_bar(stat = "identity", position = "stack") +
  coord_flip() +
  scale_fill_manual(values = c("Up_regulated" = "brown", "Down_regulated" = "lightblue")) +
  labs(
    title = "Differentially Expressed Genes by Sample Factor",
    subtitle = "Fold Change >= 2, Alpha <= 0.2",
    x = "Comparison",
    y = "Number of Differentially Expressed Genes"
  ) +
  theme_minimal(base_size = 14) +
  theme(plot.title.position = "plot")
```

#q8
the reason the samples cluster so much better now is simple: my first analysis was noisy because it included every single gene, but this heatmap only uses the genes that actually responded to the treatment. By filtering out the noise from all the unchanging genes, the real biological patterns emerge. This result makes much more sense, as the heatmap now clearly shows how the plant reacts differently to the mock, avirulent, and virulent challenges.
